golang/cosmos/x/swingset/genesis.go

```diff
@@ -46,15 +54,62 @@ func InitGenesis(ctx sdk.Context, k Keeper, swingStoreExportsHandler *SwingStore
 
 	swingStore := k.GetSwingStore(ctx)
 
-	for _, entry := range swingStoreExportData {
-		swingStore.Set([]byte(entry.Key), []byte(entry.Value))
-	}
-
 	snapshotHeight := uint64(ctx.BlockHeight())
 
-	getExportDataReader := func() (agoric.KVEntryReader, error) {
-		exportDataIterator := swingStore.Iterator(nil, nil)
-		return agoric.NewKVIteratorReader(exportDataIterator), nil
+	var getExportDataReader func() (agoric.KVEntryReader, error)
+
+	if len(swingStoreExportData) > 0 {
+		for _, entry := range swingStoreExportData {
+			swingStore.Set([]byte(entry.Key), []byte(entry.Value))
+		}
+		getExportDataReader = func() (agoric.KVEntryReader, error) {
+			exportDataIterator := swingStore.Iterator(nil, nil)
+			return agoric.NewKVIteratorReader(exportDataIterator), nil
+		}
+	} else {
+		hashParts := strings.SplitN(data.SwingStoreExportDataHash, ":", 2)
+		if len(hashParts) != 2 {
+			panic(fmt.Errorf("invalid swing-store export data hash %s", data.SwingStoreExportDataHash))
+		}
+		if hashParts[0] != "sha256" {
+			panic(fmt.Errorf("invalid swing-store export data hash algorithm %s, expected sha256", hashParts[0]))
+		}
+		sha256Hash, err := hex.DecodeString(hashParts[1])
+		if err != nil {
+			panic(err)
+		}
+		getExportDataReader = func() (agoric.KVEntryReader, error) {
+			kvReader, err := artifactProvider.GetExportDataReader()
+			if err != nil {
+				return nil, err
+			}
+
+			if kvReader == nil {
+				return nil, fmt.Errorf("swing-store export has no export data")
+			}
+
+			hasher := sha256.New()
+			encoder := json.NewEncoder(hasher)
+			encoder.SetEscapeHTML(false)
+
+			return agoric.NewKVHookingReader(kvReader, func(entry agoric.KVEntry) error {
+				key := []byte(entry.Key())
+
+				if !entry.HasValue() {
+					swingStore.Delete(key)
+				} else {
+					swingStore.Set(key, []byte(entry.StringValue()))
+				}
+
+				return encoder.Encode(entry)
```

Comment:
I went back and forth on this until I realized that there is no canonical JSON encoding. As such, hashing the original file isn't quite right. Of course even with a specified JSON encoding of each entry, there is no canonical way in which the entries should appear. In theory we should use an IAVL tree of the data to have a canonical hash :)

Basing the hash on the re-encoding to JSON by the golang encoder is thus motivated both by practicality (it fit better in the abstraction already in place of "export data reader"), and slightly increases the canonical status of the hash.

That said I was also careful that the hash computed this way is exactly equal to `sha256sum` of the file created by golang serialization.

> many-to-one Unicode normalization of some codepoint X into bytes Z1Z2, and also serializes codepoint Y into the same Z1Z2. Then this could be tricked into passing Y into the swingstore data when the creator meant (and hashed) X.

As mentioned above, it's actually the opposite. There are multiple ways to escape the same string value in JSON encodings, and here we're settling on the one that golang implements

Reply:
> there is no canonical JSON encoding
> there is no canonical way in which the entries should appear

Yeah, that's the clue that we can't/shouldn't rely upon "canonical" encoding at all. In general, the creator of the hashed/signed artifact can encode it any way they want: doesn't matter, nobody else will be attempting to encode it in the same way. What matters is that everybody's *decoder* will treat the hashed/signed bytes in the same way.

If we find ourselves wanting a canonical *encoder*, that might mean we're doing encoding on the decoding/verification side, and that's where vulnerabilities live.